File: compiler/python/state-buffer-estimate/layers/combinelayer.py
from utils.consts    import *
from utils.fmapdesc  import OfmapDesc
from .layer           import Layer
import nets.network

##########################################################
class CombineLayer(Layer): # abstract class
    #-----------------------------------------------------------------
    def __init__(self, param, prev_layer, earlier_layer, num_ofmaps):
        assert(isinstance(prev_layer, Layer))
        assert(isinstance(earlier_layer, Layer))
        assert(prev_layer.gOfmapWidth() == earlier_layer.gOfmapWidth())
        assert(prev_layer.gOfmapHeight() == earlier_layer.gOfmapHeight())

        ofmap_desc = OfmapDesc(num_ofmaps, 
                          (prev_layer.gOfmapWidth(), prev_layer.gOfmapHeight()))

        super().__init__(param, ofmap_desc, (prev_layer, earlier_layer))

    #-----------------------------------------------------------------
    def gJson(self):
        x = super().gJson()
        return x

    #-----------------------------------------------------------------
    def qPassThrough(self):
        return False

    #-----------------------------------------------------------------
    def qCombineLayer(self):
        return True

File: compiler/kcc/layers/src/combinelayer.cpp
#include "combinelayer.hpp"


namespace kcc {

namespace nets {
    class Network;
}

namespace layers {

//--------------------------------------------------------
CombineLayer::CombineLayer(const Params& params, Layer* prev_layer, Layer* earlier_layer, int num_ofmaps)
    : Layer(params,
            FmapDesc(num_ofmaps, prev_layer->gOfmapHeight(), prev_layer->gOfmapWidth()),
            prev_layer->gDataTensorDimSemantics(),
            mkLayerVector2(prev_layer, earlier_layer))

{
    assert(prev_layer->gOfmapWidth() == earlier_layer->gOfmapWidth());
    assert(prev_layer->gOfmapHeight() == earlier_layer->gOfmapHeight());
}

//----------------------------------------------------------------
bool
CombineLayer::qPassThrough() const
{
    return false;
}

//----------------------------------------------------------------
bool
CombineLayer::qCombineLayer() const
{
    return true;
}

} // namespace layers
} // namespace kcc

